### PR TITLE
feat: Add clustering support for CREATE TABLE

### PR DESCRIPTION
**Pull request by sanujbasu**
## 🥞 Stacked PR
Use this [link](https://github.com/delta-io/delta-kernel-rs/pull/1763/files) to review incremental changes.
- [**stack/create_table_simple_clustering**](https://github.com/delta-io/delta-kernel-rs/pull/1763) [[Files changed](https://github.com/delta-io/delta-kernel-rs/pull/1763/files)]
  - [stack/create_table_simple_cm](https://github.com/delta-io/delta-kernel-rs/pull/1764) [[Files changed](https://github.com/delta-io/delta-kernel-rs/pull/1764/files/a374a1a02ec34996e4f629a807966ed2f0a6065a..417bf953ee0d53d6396e838363b4248edec86054)]

---------
## What changes are proposed in this pull request?

This adds the ability to create clustered Delta tables using the
`with_data_layout(DataLayout::Clustered { columns })` API.

Key changes:
- Add DataLayout enum with None and Clustered variants
- Add with_data_layout() method to CreateTableTransactionBuilder
- Implement support_clustering_for_table_create() for validation and
  domain metadata creation
- Update Transaction::try_new_create_table() to accept clustering columns
- Allowlist TableFeature::DomainMetadata for CREATE TABLE
- Add unit tests for clustering validation
- Add integration tests for clustered table creation and stats columns

Clustering columns are validated to:
- Exist in the schema
- Be top-level (not nested)
- Have at most 4 columns (MAX_CLUSTERING_COLUMNS)
- The clustering columns are passed to Transaction so stats_columns()
correctly includes them regardless of schema position.

## How was this change tested?
Unit and integration tests

### CODE REVIEW

File: kernel/src/transaction/create_table.rs [hunk 1]
```diff
@@ -138,6 +142,69 @@ struct ValidatedTableProperties {
...
+/// - ReaderWriter features are added to both reader and writer lists
+/// - Writer and Unknown features are added only to the writer list
+fn add_feature_to_lists(
+    feature: TableFeature,
+    reader_features: &mut Vec<TableFeature>,
+    writer_features: &mut Vec<TableFeature>,
+) {
+    match feature.feature_type() {
+        FeatureType::ReaderWriter => {
+            reader_features.push(feature.clone());
```
Comment (OussamaSaoudi) [inline kernel/src/transaction/create_table.rs]: Add a check here that ensures reader/writer features doesn't already contain this. Imagine
`'delta.domainMetadata.support' = 'supported'` + dataLayout = clustering.

Comment (sanujbasu) [reply kernel/src/transaction/create_table.rs]: good catch. fixed and added a test.

File: kernel/src/transaction/create_table.rs [hunk 2]
```diff
@@ -56,9 +60,10 @@ use crate::{DeltaResult, Engine, Error, StorageHandler, PRE_COMMIT_VERSION};
 /// Feature signals (`delta.feature.X=supported`) are validated against this list.
 /// Only features in this list can be enabled via feature signals.
 ///
-/// This list will expand as more features are supported (e.g., column mapping, clustering).
+/// This list will expand as more features are supported (e.g., column mapping).
 const ALLOWED_DELTA_FEATURES: &[TableFeature] = &[
-    // Currently empty - no features allowed yet
+    // DomainMetadata is required for clustering and other system domain operations
+    TableFeature::DomainMetadata,
```
Comment (scottsand-db) [inline kernel/src/transaction/create_table.rs:66]: So no `delta.feature.clustering = supported`.

What is the standard CUJ? Do users typically set that expliciitly? Or no?

File: kernel/src/transaction/create_table.rs [hunk 3]
```diff
@@ -138,6 +143,77 @@ struct ValidatedTableProperties {
...
+/// # Returns
+///
+/// The domain metadata action containing the clustering configuration.
+fn apply_clustering_for_table_create(
+    schema: &SchemaRef,
+    columns: &[crate::expressions::ColumnName],
+    reader_features: &mut Vec<TableFeature>,
+    writer_features: &mut Vec<TableFeature>,
+) -> DeltaResult<crate::actions::DomainMetadata> {
+    // Validate columns against schema (top-level, exist in schema)
```
Comment (scottsand-db) [inline kernel/src/transaction/create_table.rs]: Must clustering columns be top-level? I don't think that's true?

File: kernel/src/transaction/create_table.rs [hunk 4]
```diff
@@ -535,4 +647,125 @@ mod tests {
...
+
+        assert!(result.is_err());
+        assert!(result
+            .unwrap_err()
+            .to_string()
+            .contains("Clustering column 'nonexistent' not found in schema"));
+    }
+
+    #[test]
+    fn test_clustering_nested_column_rejected() {
```
Comment (scottsand-db) [inline kernel/src/transaction/create_table.rs:756]: I don't think this is a requirement btw

File: kernel/src/clustering.rs [hunk 5]
```diff
@@ -34,7 +37,65 @@ struct ClusteringDomainMetadata {
 }
 
 /// The domain name for clustering metadata.
-const CLUSTERING_DOMAIN_NAME: &str = "delta.clustering";
+pub(crate) const CLUSTERING_DOMAIN_NAME: &str = "delta.clustering";
+
+/// Maximum number of columns that can be used for clustering.
+///
+/// This limit matches the Delta protocol specification.
+pub(crate) const MAX_CLUSTERING_COLUMNS: usize = 4;
```
Comment (scottsand-db) [inline kernel/src/clustering.rs:47]: This isn't required per the protocol

Delta-Spark has a SQL configuration for this, but that is a connector-specific opinion, not something kernel should enforce

```
  From DeltaSQLConf.scala:2708-2718:
  val DELTA_NUM_CLUSTERING_COLUMNS_LIMIT =
    buildStaticConf("clusteredTable.numClusteringColumnsLimit")
      .internal()
      .doc("""The maximum number of clustering columns allowed for a clustered table.""")
      .intConf
      .checkValue(_ > 0, ...)
      .createWithDefault(4)
```

File: kernel/src/transaction/data_layout.rs [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+/// Data layout configuration for a Delta table.
+///
+/// Determines how data files are organized within the table:
+///
+/// - [`DataLayout::None`]: No special organization (default)
+/// - [`DataLayout::Clustered`]: Data files optimized for queries on clustering columns
+#[derive(Debug, Clone, Default)]
+pub enum DataLayout {
+    /// No special data organization (default).
+    #[default]
```
Comment (scottsand-db) [inline kernel/src/transaction/data_layout.rs:27]: i wonder if this is helpful - or if we would rather callers explicitly decide and know that they are getting None -- wdyt?

File: kernel/src/transaction/data_layout.rs [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+use crate::clustering::MAX_CLUSTERING_COLUMNS;
+use crate::expressions::ColumnName;
+use crate::{DeltaResult, Error};
+
+/// Data layout configuration for a Delta table.
+///
+/// Determines how data files are organized within the table:
+///
+/// - [`DataLayout::None`]: No special organization (default)
+/// - [`DataLayout::Clustered`]: Data files optimized for queries on clustering columns
```
Comment (scottsand-db) [inline kernel/src/transaction/data_layout.rs:21]: Add a TODO: partitioned ?

File: kernel/src/transaction/data_layout.rs [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+#[derive(Debug, Clone, Default)]
+pub enum DataLayout {
+    /// No special data organization (default).
+    #[default]
+    None,
+
+    /// Data files optimized for queries on clustering columns.
+    ///
+    /// Clustering columns must be top-level columns in the schema.
+    /// Maximum of 4 columns allowed.
```
Comment (scottsand-db) [inline kernel/src/transaction/data_layout.rs:33]: Just commenting so you know this is another place to update

File: kernel/src/transaction/create_table.rs (hunk 2 above)
Comment (sanujbasu) [reply kernel/src/transaction/create_table.rs:66]: No

In Spark you use CLUSTER BY (col1, col2) during table creation, which atomically:
Enables the ClusteredTable writer feature
Enables the DomainMetadata writer feature
Writes the delta.clustering domain metadata with the column definitions
No way to enable feature without columns.

 The feature is only enabled through the CLUSTER BY clause, which always requires columns.

The protocol doesn't explicitly forbid enabling the feature without columns, but it creates an inconsistent state. In practice, one should always define clustering columns when enabling the feature.

Comment (scottsand-db) [reply kernel/src/transaction/create_table.rs:66]: Sg! Makes sense

File: kernel/src/clustering.rs (hunk 5 above)
Comment (sanujbasu) [reply kernel/src/clustering.rs:47]: check this https://docs.delta.io/delta-clustering/#limitations.

File: kernel/src/transaction/create_table.rs (hunk 3 above)
Comment (OussamaSaoudi) [reply kernel/src/transaction/create_table.rs]: you're right -- I'm not seeing that in [clustering](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#clustered-table)

Further, seems that this works (Note clustered by `address.city`):
```sql
CREATE TABLE clustered_nested_demo (
  customer_id INT,
  customer_name STRING,
  address STRUCT<street: STRING, city: STRING, state: STRING, zip: STRING>,
  order_date DATE,
  order_amount DECIMAL(10,2)
) CLUSTER BY (address.city, order_date)
```

Comment (sanujbasu) [reply kernel/src/transaction/create_table.rs]: Good catch @scottsand-db. There seems to be some confusion amongst the different engines.
Filed https://github.com/delta-io/delta-kernel-rs/issues/1794 as discussed. Thanks @OussamaSaoudi for validating.

File: kernel/src/transaction/create_table.rs (hunk 4 above)
Comment (sanujbasu) [reply kernel/src/transaction/create_table.rs:756]: https://github.com/delta-io/delta-kernel-rs/issues/1794

File: kernel/src/clustering.rs (hunk 5 above)
Comment (sanujbasu) [reply kernel/src/clustering.rs:47]: https://github.com/delta-io/delta-kernel-rs/issues/1794

File: kernel/src/transaction/data_layout.rs (hunk 8 above)
Comment (sanujbasu) [reply kernel/src/transaction/data_layout.rs:33]: Ack. https://github.com/delta-io/delta-kernel-rs/issues/1794

File: kernel/src/transaction/data_layout.rs (hunk 7 above)
Comment (sanujbasu) [reply kernel/src/transaction/data_layout.rs:21]: done

File: kernel/src/transaction/data_layout.rs (hunk 6 above)
Comment (sanujbasu) [reply kernel/src/transaction/data_layout.rs:27]: I can consider removing None when i add in the partitioning type. Currently, it will be weird to have a single variant enum :P

File: kernel/src/transaction/create_table.rs [hunk 9]
```diff
@@ -56,9 +60,10 @@ use crate::{DeltaResult, Engine, Error, StorageHandler, PRE_COMMIT_VERSION};
 /// Feature signals (`delta.feature.X=supported`) are validated against this list.
 /// Only features in this list can be enabled via feature signals.
 ///
-/// This list will expand as more features are supported (e.g., column mapping, clustering).
+/// This list will expand as more features are supported (e.g., column mapping).
 const ALLOWED_DELTA_FEATURES: &[TableFeature] = &[
```
Comment (scottsand-db) [inline kernel/src/transaction/create_table.rs:64]: `// Note: Clustering should never be included here. Users are not expected to enable clustering via `delta.feature.clustering = supported`

File: kernel/src/transaction/create_table.rs [hunk 10]
```diff
@@ -138,6 +143,77 @@ struct ValidatedTableProperties {
...
+/// 1. Validates clustering columns against the schema (via `clustering::validate_clustering_columns`)
+/// 2. Adds DomainMetadata and ClusteredTable features to the protocol
+/// 3. Creates the domain metadata action (via `clustering::create_clustering_domain_metadata`)
+///
+/// # Arguments
+///
+/// * `schema` - The table schema to validate columns against
+/// * `columns` - The clustering column names
+/// * `reader_features` - Reader features list to update
+/// * `writer_features` - Writer features list to update
```
Comment (scottsand-db) [inline kernel/src/transaction/create_table.rs]: nit: imo if we just have self-describing names, we can get tid of this comment bloat. e.g. rename schema to table_schema, rename columns to clustering_columns, and then we are good.

i defer to you. even if you don't delete this argument docs, can you at least rename the `clustering_columns` variable

File: kernel/src/transaction/create_table.rs [hunk 11]
```diff
@@ -138,6 +143,77 @@ struct ValidatedTableProperties {
...
+///
+/// # Returns
+///
+/// The domain metadata action containing the clustering configuration.
+fn apply_clustering_for_table_create(
+    schema: &SchemaRef,
+    columns: &[crate::expressions::ColumnName],
```
Comment (scottsand-db) [inline kernel/src/transaction/create_table.rs]: physical or logical?

File: kernel/src/transaction/create_table.rs [hunk 12]
```diff
@@ -539,4 +651,125 @@ mod tests {
             "Setting delta property 'delta.appendOnly' is not supported",
         );
     }
+
+    #[test]
+    fn test_clustering_support_valid() {
```
Comment (scottsand-db) [inline kernel/src/transaction/create_table.rs:671]: test: cannot set `delta.feature.clustering = supported`

File: kernel/src/transaction/create_table.rs [hunk 13]
```diff
@@ -539,4 +651,125 @@ mod tests {
...
+        use crate::clustering::CLUSTERING_DOMAIN_NAME;
```
Comment (scottsand-db) [inline kernel/src/transaction/create_table.rs:672]: also test: that we *can* set `delta.feature.domainMetadata = supported` // right? we can do that now?

File: kernel/tests/create_table.rs [hunk 14]
```diff
@@ -263,3 +263,245 @@ async fn test_create_table_log_actions() -> DeltaResult<()> {
...
+    let table_path = temp_dir.path().to_str().expect("Invalid path").to_string();
+
+    // Build schema with 40 columns
+    let fields: Vec<StructField> = (0..40)
+        .map(|i| StructField::new(format!("col{}", i), DataType::INTEGER, true))
+        .collect();
+    let schema = Arc::new(StructType::try_new(fields)?);
+
+    let engine =
+        create_default_engine(&url::Url::from_directory_path(&table_path).expect("Invalid URL"))?;
```
Comment (scottsand-db) [inline kernel/tests/create_table.rs]: We see this a few times. can we just create a `fn create_test_engine()`?

File: kernel/tests/create_table.rs [hunk 15]
```diff
@@ -263,3 +263,245 @@ async fn test_create_table_log_actions() -> DeltaResult<()> {
...
+///
+/// This tests the edge case where a user provides `delta.feature.domainMetadata=supported`
+/// AND uses `DataLayout::Clustered`. Both would try to add DomainMetadata, but we should
+/// only have it once in the feature lists.
+#[tokio::test]
+async fn test_clustering_with_explicit_feature_signal_no_duplicates() -> DeltaResult<()> {
+    use delta_kernel::transaction::data_layout::DataLayout;
+
+    let temp_dir = tempdir().expect("Failed to create temp dir");
+    let table_path = temp_dir.path().to_str().expect("Invalid path").to_string();
```
Comment (scottsand-db) [inline kernel/tests/create_table.rs]: i wonder if we should have a test setup helper taht returns temp_dir, table_path, and engine? all in 1 line?

`let { temp_dir, table_path, engine } = setup()`

File: kernel/src/transaction/create_table.rs [hunk 16]
```diff
@@ -539,4 +651,125 @@ mod tests {
...
+            &[ColumnName::new(["id"]), ColumnName::new(["date"])],
+            &mut reader_features,
+            &mut writer_features,
+        )
+        .unwrap();
+
+        // Verify domain metadata contains both columns
+        let config: serde_json::Value = serde_json::from_str(dm.configuration()).unwrap();
+        let clustering_cols = config["clusteringColumns"].as_array().unwrap();
+        assert_eq!(clustering_cols.len(), 2);
```
Comment (OussamaSaoudi) [inline kernel/src/transaction/create_table.rs:723]: We didn't check that the clustering columns are id and date.

File: kernel/src/clustering.rs [hunk 17]
```diff
@@ -34,7 +37,68 @@ struct ClusteringDomainMetadata {
...
+                col
+            )));
+        }
+        let col_name = &col.path()[0];
+        if schema.field(col_name).is_none() {
+            return Err(Error::generic(format!(
+                "Clustering column '{}' not found in schema",
+                col_name
+            )));
+        }
```
Comment (OussamaSaoudi) [inline kernel/src/clustering.rs:133]: I think we need to validate the clustering type. From [the docs](https://docs.databricks.com/aws/en/delta/clustering#supported-data-types):

> Clustering supports these data types for clustering keys:
> Date
> Timestamp
> TimestampNTZ (Databricks Runtime 14.3 LTS and above)
> String
> Integer, Long, Short, Byte
> Float, Double, Decimal

Comment (OussamaSaoudi) [reply kernel/src/clustering.rs:133]: Can you add tests for these types ^? Also ensure that other types aren't allowed?

File: kernel/tests/create_table.rs (hunk 15 above)
Comment (OussamaSaoudi) [reply kernel/tests/create_table.rs]: I think that exists in test-utils.

File: kernel/tests/create_table.rs [hunk 18]
```diff
@@ -263,3 +263,245 @@ async fn test_create_table_log_actions() -> DeltaResult<()> {
...
+    // Read back and verify no duplicate features
+    let commit_file = std::fs::read_to_string(format!(
+        "{}/_delta_log/00000000000000000000.json",
+        table_path
+    ))?;
+
+    let actions: Vec<Value> = commit_file
+        .lines()
+        .map(|l| serde_json::from_str(l).unwrap())
+        .collect();
```
Comment (OussamaSaoudi) [inline kernel/tests/create_table.rs]: can you please use kernel's APIs to read this? It's a delta table after all, and we perform a bunch of checks upon read. That will help make this test more likely to catch errors/bugs.

File: kernel/src/transaction/create_table.rs (hunk 9 above)
Comment (sanujbasu) [reply kernel/src/transaction/create_table.rs:64]: Added

File: kernel/src/transaction/create_table.rs (hunk 10 above)
Comment (sanujbasu) [reply kernel/src/transaction/create_table.rs]: done

File: kernel/src/transaction/create_table.rs (hunk 12 above)
Comment (sanujbasu) [reply kernel/src/transaction/create_table.rs:671]: Extended test_validate_unsupported_properties

File: kernel/src/transaction/create_table.rs (hunk 13 above)
Comment (sanujbasu) [reply kernel/src/transaction/create_table.rs:672]: Extended test_validate_supported_properties to verify delta.feature.domainMetadata=supported is allowed

File: kernel/src/transaction/create_table.rs (hunk 16 above)
Comment (sanujbasu) [reply kernel/src/transaction/create_table.rs:723]: done

File: kernel/src/clustering.rs (hunk 17 above)
Comment (sanujbasu) [reply kernel/src/clustering.rs:133]: done

File: kernel/tests/create_table.rs (hunk 15 above)
Comment (sanujbasu) [reply kernel/tests/create_table.rs]: Didnt find one in test-utils. Created one

File: kernel/tests/create_table.rs (hunk 18 above)
Comment (sanujbasu) [reply kernel/tests/create_table.rs]: Using snapshot to read back now

File: kernel/tests/create_table.rs (hunk 14 above)
Comment (sanujbasu) [reply kernel/tests/create_table.rs]: created

File: kernel/src/transaction/create_table.rs [hunk 19]
```diff
@@ -56,9 +60,13 @@ use crate::{DeltaResult, Engine, Error, StorageHandler, PRE_COMMIT_VERSION};
...
 ///
-/// This list will expand as more features are supported (e.g., column mapping, clustering).
+/// This list will expand as more features are supported (e.g., column mapping).
 const ALLOWED_DELTA_FEATURES: &[TableFeature] = &[
-    // Currently empty - no features allowed yet
+    // DomainMetadata is required for clustering and other system domain operations
+    TableFeature::DomainMetadata,
+    // Note: Clustering is NOT included here. Users should not enable clustering via
+    // `delta.feature.clustering = supported`. Instead, clustering is enabled by
+    // specifying clustering columns via `with_clustering_columns()`.
```
Comment (scottsand-db) [inline kernel/src/transaction/create_table.rs]: isn't it `.with_data_layout`?

File: kernel/src/transaction/data_layout.rs [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+                "Clustered layout supports at most {} columns, got {}",
+                MAX_CLUSTERING_COLUMNS,
+                columns.len()
+            )));
+        }
+
+        // Check for duplicate columns
+        let mut seen = HashSet::new();
+        for col in &columns {
+            let col_name = &col.path()[0];
```
Comment (OussamaSaoudi) [inline kernel/src/transaction/data_layout.rs]: We're assuming in this piece of code that path is of length 1. Let's ensure that.
```suggestion
            // TODO(...): We only support clustering on top-level columns.
            require!(
                col.path().len() == 1,
                Error::generic(format!("Cannot apply clustering to the column {}. Only top-level columns are currently supported.", col))
            );
            let col_name = &col.path()[0];
```

Comment (OussamaSaoudi) [reply kernel/src/transaction/data_layout.rs]: Add a test below pls.

File: kernel/src/transaction/data_layout.rs [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+    fn test_clustered_layout_valid() {
+        let layout = DataLayout::clustered(["col1", "col2"]).unwrap();
+        assert!(layout.is_clustered());
+        if let DataLayout::Clustered { columns } = layout {
+            assert_eq!(columns.len(), 2);
+        }
+    }
+
+    #[test]
+    fn test_clustered_layout_empty_columns() {
```
Comment (OussamaSaoudi) [inline kernel/src/transaction/data_layout.rs]: For future reference: I recommend using [rstest](https://docs.rs/rstest/latest/rstest/) :) We recently added it to kernel 🎉

File: kernel/src/clustering.rs [hunk 22]
```diff
@@ -27,14 +31,94 @@ use crate::{DeltaResult, Engine};
...
+pub(crate) const MAX_CLUSTERING_COLUMNS: usize = 4;
+
+/// Validates clustering columns against the table schema.
+///
+/// This function validates that:
+/// 1. Columns are top-level (not nested paths)
+/// 2. Columns exist in the schema
+/// 3. Columns have data types eligible for statistics collection (required for clustering)
+///
+/// Note: The number of columns (1-4) should be validated before calling this function,
```
Comment (scottsand-db) [inline kernel/src/clustering.rs]: Why this? What if we moved that logic here?

File: kernel/tests/create_table.rs [hunk 23]
```diff
@@ -11,17 +11,11 @@ use delta_kernel::table_features::{
 use delta_kernel::transaction::create_table::create_table;
 use delta_kernel::DeltaResult;
 use serde_json::Value;
-use tempfile::tempdir;
-use test_utils::{assert_result_error_with_message, create_default_engine};
+use test_utils::{assert_result_error_with_message, test_table_setup};
 
 #[tokio::test]
 async fn test_create_simple_table() -> DeltaResult<()> {
```
Comment (scottsand-db) [inline kernel/tests/create_table.rs:25]: I think we need tests for user domain metadata now, correct?

That is, an e2e
(1) invoke create table builder
(2) build, get a txn
(3) txn.with_domain_metadata
(4) commit
(5) load back and read that user domain metadata

File: kernel/tests/create_table.rs [hunk 24]
```diff
@@ -263,3 +231,196 @@ async fn test_create_table_log_actions() -> DeltaResult<()> {
...
+    // Create schema for a clustered table
+    let schema = Arc::new(StructType::try_new(vec![
+        StructField::new("id", DataType::INTEGER, false),
+        StructField::new("name", DataType::STRING, true),
+        StructField::new("timestamp", DataType::TIMESTAMP, false),
+    ])?);
+
+    // Create clustered table on "id" column
+    let txn = create_table(&table_path, schema.clone(), "DeltaKernel-RS/Test")
+        .with_data_layout(DataLayout::clustered(["id"])?)
```
Comment (scottsand-db) [inline kernel/tests/create_table.rs]: Do we need any tests for: schema has `id` but I pass in `ID` ? what's the expected behavior there?

File: kernel/tests/create_table.rs [hunk 25]
```diff
@@ -263,3 +231,196 @@ async fn test_create_table_log_actions() -> DeltaResult<()> {
...
+
+    // Verify stats_columns includes the clustering column
+    let stats_cols = txn.stats_columns();
+    assert!(
+        stats_cols.iter().any(|c| c.to_string() == "id"),
+        "Clustering column 'id' should be in stats columns"
+    );
+
+    // Commit the table
+    let _result = txn.commit(engine.as_ref())?;
```
Comment (scottsand-db) [inline kernel/tests/create_table.rs]: nit: do we need to do `let _result = txn.commit(..)` or can we just do `txn.commit(...)` ?

File: kernel/src/clustering.rs [hunk 26]
```diff
@@ -101,4 +186,154 @@ mod tests {
         let expected_cols: Vec<ColumnName> = expected.into_iter().map(ColumnName::new).collect();
         assert_eq!(columns, expected_cols);
     }
+
```
Comment (OussamaSaoudi) [inline kernel/src/clustering.rs:227]: Can you add a test case with an empty ColumnName?

File: kernel/tests/create_table.rs [hunk 27]
```diff
@@ -263,3 +231,196 @@ async fn test_create_table_log_actions() -> DeltaResult<()> {
...
+        .writer_features()
+        .expect("Writer features should exist");
+    assert!(
+        writer_features.contains(&TableFeature::DomainMetadata),
+        "Protocol should have domainMetadata writer feature"
+    );
+    assert!(
+        writer_features.contains(&TableFeature::ClusteredTable),
+        "Protocol should have clustering writer feature"
+    );
```
Comment (OussamaSaoudi) [inline kernel/tests/create_table.rs]: pls use
```rust
    /// Generic method to check if a feature is supported in the protocol.
    /// This does NOT check if the feature is enabled via table properties.
    #[internal_api]
    pub(crate) fn is_feature_supported(&self, feature: &TableFeature) -> bool {
        let Some(info) = feature.info() else {
            return false;
        };
        self.is_feature_info_supported(feature, info)
    }

    /// Generic method to check if a feature is enabled.
    ///
    /// A feature is enabled if:
    /// 1. It is supported in the protocol
    /// 2. The enablement check passes
    #[internal_api]
    pub(crate) fn is_feature_enabled(&self, feature: &TableFeature) -> bool {
        let Some(info) = feature.info() else {
            return false;
        };
        self.is_feature_info_enabled(feature, info)
    }
}
```

Comment (OussamaSaoudi) [reply kernel/tests/create_table.rs]:
```suggestion
    let table_configuration = snapshot.table_configuration();
    assert!(table_configuration.is_feature_enabled(&TableFeature::DomainMetadata));
    assert!(table_configuration.is_feature_enabled(&TableConfiguration::TableFeature::ClusteredTable);
    assert!(table_configuration.is_feature_supported(&TableFeature::DomainMetadata));
    assert!(table_configuration.is_feature_supported(&TableConfiguration::TableFeature::ClusteredTable);
```

File: kernel/src/transaction/create_table.rs (hunk 19 above)
Comment (sanujbasu) [reply kernel/src/transaction/create_table.rs]: Yes :) . Fixed.

File: kernel/src/clustering.rs (hunk 22 above)
Comment (sanujbasu) [reply kernel/src/clustering.rs]: `DataLayout::clustered()` validates structural constraints specifically column inputs => Do the current set of columns meet column level constraints (the schema is not available then)?

`validate_clustering_columns` is called when build is triggered which has all the schema information to validate against the columns => Can the proposed columns be applied to the schema? I don't have a strict preference in this case. I can add all the validation here.

File: kernel/tests/create_table.rs (hunk 27 above)
Comment (sanujbasu) [reply kernel/tests/create_table.rs]: Good suggestion. done

File: kernel/tests/create_table.rs (hunk 25 above)
Comment (sanujbasu) [reply kernel/tests/create_table.rs]: Update: CommitResult is marked #[must_use], which means we need to explicitly acknowledge we're ignoring it with let _ = ....

File: kernel/src/clustering.rs (hunk 22 above)
Comment (scottsand-db) [reply kernel/src/clustering.rs]: I'd say if we _can_ validate earlier (e.g. here we just want to validate this input columns against our static const of 4) then we might as well

Especially as this reduces a dependency on this function (we longer have to say that callers must do X before calling us)

Comment (sanujbasu) [reply kernel/src/clustering.rs]: Decided to move all validations to `validate_clustering_columns`

File: kernel/src/transaction/data_layout.rs (hunk 20 above)
Comment (sanujbasu) [reply kernel/src/transaction/data_layout.rs]:
```
    #[test]
    fn test_validate_clustering_columns_nested_rejected() {
        let schema =
            StructType::new_unchecked(vec![StructField::new("nested", DataType::STRING, true)]);
        let columns = vec![ColumnName::new(["nested", "field"])];
        let result = validate_clustering_columns(&schema, &columns);
        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("must be a top-level column"));
    }
```
already added

File: kernel/src/transaction/data_layout.rs (hunk 21 above)
Comment (sanujbasu) [reply kernel/src/transaction/data_layout.rs]: ooooo nice !!!! The test bloat is indeed annoying.

File: kernel/src/clustering.rs (hunk 26 above)
Comment (sanujbasu) [reply kernel/src/clustering.rs:227]: Addressed in

```
    fn test_validate_clustering_columns_empty_name_rejected() {
        let schema =
            StructType::new_unchecked(vec![StructField::new("id", DataType::INTEGER, false)]);
        // Create a ColumnName with empty path
        let columns: Vec<ColumnName> = vec![ColumnName::new(Vec::<String>::new())];
        let result = validate_clustering_columns(&schema, &columns);
        assert!(result.is_err());
        assert!(result.unwrap_err().to_string().contains("cannot be empty"));
    }

```

File: kernel/src/transaction/data_layout.rs (hunk 21 above)
Comment (sanujbasu) [reply kernel/src/transaction/data_layout.rs]: Used some here and filed https://github.com/delta-io/delta-kernel-rs/issues/1804 to look for more simplification opportunitiies.

File: kernel/tests/create_table.rs (hunk 23 above)
Comment (sanujbasu) [reply kernel/tests/create_table.rs:25]: Added `test_create_table_with_user_domain_metadata`

File: kernel/tests/create_table.rs (hunk 24 above)
Comment (sanujbasu) [reply kernel/tests/create_table.rs]: This is good point! Based on what I find there seems to be a discrepancy b/w the Kernel and Spark. Spark is case insensitive while specifying the "id"/ "iD" / "ID" columns but the Kernel seems to be case sensitive.
Filed #1805  for tracking.

File: kernel/tests/create_table.rs [hunk 28]
```diff
@@ -88,14 +79,68 @@ async fn test_create_simple_table() -> DeltaResult<()> {
     Ok(())
 }
 
+/// Test adding user domain metadata during table creation.
+///
+/// This tests the flow:
+/// 1. create_table() builder → build() → Transaction
+/// 2. txn.with_domain_metadata()
+/// 3. commit()
+/// 4. Load snapshot and verify via get_domain_metadata()
```
Comment (scottsand-db) [inline kernel/tests/create_table.rs]: nit: IMO, code is clean! no need for comment. if we keep this, when we update the code (which already has inline comments) we will have to remember to update this, too

Comment (sanujbasu) [reply kernel/tests/create_table.rs]: Cleaned up the comment
